src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -7944,37 +7944,56 @@ int RGWRados::Object::Read::iterate(const DoutPrefixProvider *dpp, int64_t ofs,
   auto aio = rgw::make_throttle(window_size, y);
   get_obj_data data(store, cb, &*aio, ofs, y);
 
-  int r = store->iterate_obj(dpp, source->get_ctx(), source->get_bucket_info(), state.obj,
-                             ofs, end, chunk_size, _get_obj_iterate_cb, &data, y);
-  if (r < 0) {
-    ldpp_dout(dpp, 0) << "iterate_obj() failed with " << r << dendl;
-    data.cancel(); // drain completions without writing back to client
-    return r;
+  RGWObjectCtx& obj_ctx = source->get_ctx();
+  RGWBucketInfo& bucket_info = source->get_bucket_info();
+
+  RGWObjState *astate = NULL;
+  RGWObjManifest *manifest = nullptr;
+
+  int r = 0;
+
+  static constexpr auto MAX_ECANCELED_RETRY = 20u;
```

Comment:
what motivates 20 retries, specifically?

Reply:
no particular reason. just feel that it's a reasonable number.
the number of retries is 100 for the [set_olh](https://github.com/ceph/ceph/blob/main/src/rgw/driver/rados/rgw_rados.cc#L9071), I feel 100 may be too many here, and the request may be timed out.
any suggestions?

Reply:
if retrying is the right response, I like 20 a lot better than 100;  I think my main question is whether the cancellation case could have been avoided--I mean, iiuc, all the racing ops are GET, so are idempotent

Reply:
In this particular case, the cancellation happens due to the `id_tag` mismatch caused by the racing replication_status_header change. This can be fix in `handle_replication_status_header` by reloading the object attributes, and without the need for retry in `RGWRados::Object::Read::iterate`.

However there are other possible conflicts, such as pending olh attribute removal, or concurrent GET and PUT etc.
I do see some GET returns 409 while there is a concurrent PUT in our production logs.

So I think we might as well do a blanket retry on -ECANCELED on read operation. If I understand the -ECANCELED correctly, it should be a racing conflict that can be recovered by reload and retry.

Reply:
ok, I buy that for this, but removing avoidable canceled cycles as well should improve performance

Reply:
as you mentioned in https://tracker.ceph.com/issues/70992:
> this error was sent back to the client with code 500 until { ECANCELED, {409, "ConcurrentModification"}} is added into rgw_http_s3_errors in this commit: https://github.com/ceph/ceph/commit/a9c49a5ce7a2

prior to that, we were returning a 500 error so retries were driven by the client. a retry or two here seems reasonable, but if we _do_ give up we should map the `ECANCELED` to something like `ERR_INTERNAL_ERROR` so the client will try again